### DISCUSSION TITLE
Help modifyin/adding new nodes for a custom implementation of glsp (node server) integrated with theia

### DISCUSSION THREAD

**Discussion by Mansalek** [eclipse-glsp/glsp#615] [Q&A]
Good morning ,
i'am looking to make a custom implementation of the Workflow example (since there is no minimal example with a 'nodejs server'  as opposed to the java one ) , where i'll be changing the default existing nodes into custom ones with different names and different behaviors(A only linked with C , different views etc)  , while keeping all of the functionalities that already exist but I find my self being overwhelmed with the workflow example being big and having it's JSON file  containing the GModel directly.
I'am basically just looking to link a json file containing my elements with the existing implementation , I have tried with the documentation but it somehow feels lacking .
I know this might seem simple so thats why it would be great if someone could guide me , thank you .

**Comment by planger** [comment]
Hi,

@tortmayr is currently working on providing a template project for this or at least a very similar set up (nodejs, custom JSON format, VS Code integration). So if you can wait until end of next week, it should be easier to wait.

Essentially, instead of extending the `GModelDiagramModule` you'd need to extend the `DiagramModule` directly and bind at least your custom source model storage, a custom model state to store your custom source model in the model state, and your GModel factory to translate your source model into the GModel. Similar to how it is described in the docs for [model loading](https://www.eclipse.org/glsp/documentation/modelloading/) and [GModel generation](https://www.eclipse.org/glsp/documentation/modelgeneration/).

```typescript
bind(SourceModelStorage).to(YourModelStorage);
bind(YourModelState).toSelf().inSingletonScope();
bind(ModelState).toService(YourModelState);
bind(GModelFactory).to(YourModelFactory).inSingletonScope();
```

Note however that this is the API as currently defined in master. The latest release diverges a bit from that (the SourceModelStorage was called `ModelSourceLoader`.

**Reply by Mansalek** [reply]
> Hi,
> 
> @tortmayr is currently working on providing a template project for this or at least a very similar set up (nodejs, custom JSON format, VS Code integration). So if you can wait until end of next week, it should be easier to wait.
> 
> Essentially, instead of extending the `GModelDiagramModule` you'd need to extend the `DiagramModule` directly and bind at least your custom source model storage, a custom model state to store your custom source model in the model state, and your GModel factory to translate your source model into the GModel. Similar to how it is described in the docs for [model loading](https://www.eclipse.org/glsp/documentation/modelloading/) and [GModel generation](https://www.eclipse.org/glsp/documentation/modelgeneration/).
> 
> ```ts
> bind(SourceModelStorage).to(YourModelStorage);
> bind(YourModelState).toSelf().inSingletonScope();
> bind(ModelState).toService(YourModelState);
> bind(GModelFactory).to(YourModelFactory).inSingletonScope();
> ```
> 
> Note however that this is the API as currently defined in master. The latest release diverges a bit from that (the SourceModelStorage was called `ModelSourceLoader`.

Hello ,
That's great! i will be awaiting @tortmayr 's template project then , surely it will clear up any confusion i had
Thank you

**Reply by tortmayr** [reply]
@Mansalek 
The new project templates are now available and the glsp [documentaion](https://www.eclipse.org/glsp/documentation/gettingstarted/)
has been updated as well. You can find the template for node, custom json and VS code here: https://github.com/eclipse-glsp/glsp-examples/tree/master/project-templates/node-json-vscode

**Comment by rsoika** [comment]
@planger: I have also just started a view days ago to implement my first custom implementation loading my model. 
I now switched from glsp-server-version 0.9.0 to 0.10.0-SNAPSHOT so I can follow the latest [docu](https://www.eclipse.org/glsp/documentation/sourcemodel/)

In this way I think I was able to implement my custom `SourceModelStorage` with the new  interface. (I used the ModelSourceLoader before)

But your binding example above is confusing me now. I have extended the `DiagramModule` and I have simply overwritten the methods `bindSourceModelStorage` and `bindGModelFactory`

```java
public class BPMNDiagramModule extends DiagramModule {
    @Override
    protected Class<? extends SourceModelStorage> bindSourceModelStorage() {
        return BPMNSourceModelStorage.class;
    }

    @Override
    protected Class<? extends GModelFactory> bindGModelFactory() {
        return BPMNGModelFactory.class;
    }
 ....
```

I do not understand what you mean with the example code

```java
bind(SourceModelStorage).to(YourModelStorage);
bind(YourModelState).toSelf().inSingletonScope();
bind(ModelState).toService(YourModelState);
bind(GModelFactory).to(YourModelFactory).inSingletonScope();
```

Does this mean we should not overwrite the `bindSourceModelStorage` and `bindGModelFactory` methods of the `DiagramModule` and instead do the binding somewhere else?

Also my ServerLauncher seems now complaining about the line:

     ServerModule serverModule = new ServerModule().configureDiagramModule(new BPMNDiagramModule());

The method expect now a mixinModules?

**Reply by tortmayr** [reply]
@rsoika The example code above was referring to the Node GLSP Server implementation. Since you are using the Java GLSP server it is not relevant for you. 
Overriding the `bindSourceModelStorage` and `bindGmodelFactory` methods is the correct approach. It might be helpful if you have a look at the new project template for the `Java Server/EMF model/Theia` usecase: https://github.com/eclipse-glsp/glsp-examples/tree/master/project-templates/java-emf-theia

> Also my ServerLauncher seems now complaining about the line:
 ServerModule serverModule = new ServerModule().configureDiagramModule(new BPMNDiagramModule())
 
 What is the error message you receive. You code looks fine to me and it is done in the same way in corresponding [project template](https://github.com/eclipse-glsp/glsp-examples/blob/master/project-templates/java-emf-theia/glsp-server/src/main/java/org/eclipse/glsp/example/javaemf/launch/TaskListServerLauncher.java#L39)

The mixin modules are optional and allows you to load additional modules on top of the `DiagramModule`

**Reply by rsoika** [reply]
@tortmayr  Yes the new code is now nearly perfectly running. The only problem that still bothers me is the GModelIndex.

After all my server complains not to find any elements in the index
```
org.eclipse.glsp.server.types.GLSPServerException: Model element not found! ID: 1457e0d2-7587-44c3-86f4-97a8e3fb0cf2_header
	at org.eclipse.glsp.server.types.GLSPServerException.getOrThrow(GLSPServerException.java:58)
	at org.eclipse.glsp.server.utils.LayoutUtil.applyBounds(LayoutUtil.java:54)
	at org.eclipse.glsp.server.features.core.model.ComputedBoundsActionHandler.executeAction(ComputedBoundsActionHandler.java:42)
	at org.eclipse.glsp.server.features.core.model.ComputedBoundsActionHandler.executeAction(ComputedBoundsActionHandler.java:28)
	at org.eclipse.glsp.server.actions.AbstractActionHandler.execute(AbstractActionHandler.java:53)
	at org.eclipse.glsp.server.internal.actions.DefaultActionDispatcher.runAction(DefaultActionDispatcher.java:191)
	at org.eclipse.glsp.server.internal.actions.DefaultActionDispatcher.handleAction(DefaultActionDispatcher.java:170)
	at org.eclipse.glsp.server.internal.actions.DefaultActionDispatcher.handleNextAction(DefaultActionDispatcher.java:157)
	at org.eclipse.glsp.server.internal.actions.DefaultActionDispatcher.runThread(DefaultActionDispatcher.java:144)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.util.NoSuchElementException: No value present
	at java.base/java.util.Optional.get(Optional.java:148)
	at org.eclipse.glsp.server.types.GLSPServerException.getOrThrow(GLSPServerException.java:56)
	... 9 more
```

This maybe makes sense as it seems I do not have such an index. But I do not understand how and why I need to implement a `GModelIndex` by myself? My custom `GModelFactory` is now able to build a `GGraph` from my BPMN model and I can successful call  `modelState.setRoot(newGModel)`

```java
public class BPMNGModelFactory implements GModelFactory {
    @Inject
    protected BPMNGModelState modelState;
    @Override
    public void createGModel() {
        BPMNModel model = modelState.getBpmnModel();
        if (model != null) {
            List<org.openbpmn.bpmn.elements.Process> processList = model.getProcesList();
            if (processList != null && processList.size() > 0) {
                org.openbpmn.bpmn.elements.Process process = processList.get(0);
                GGraph newGModel = BPMNGModelUtil.createGModelFromProcess(process, modelState);
                modelState.setRoot(newGModel);
                modelState.getRoot().setRevision(-1);
            }
        } 
    }
}
```

Is it true that I have to implement the GModelndex and traverse through the GModel build by my GModelFactory just to answer questions like where is the root or how to find a element? Should not the GModelndexImpl do the trick? And where should I register the Index?

In the moment I am totally lost.

**Reply by tortmayr** [reply]
The `GModelndex` is created automatically and should be updated when you invoke the `modelState.setRoot(newGModel)` method. There is no need for you to implement your on index version.
I had a quick look at the recent changes and it seems tha somebody introduced a breaking change with this commit :https://github.com/eclipse-glsp/glsp-server/commit/ed2040f0ee10bcf0358b20a59f78dfd2911f49f3

For no apparent reason the implementation of the original `setRoot` method in the `DefaultGModelState´ got moved to `updateRoot` and the `setRoot` method does no longer update the index. The documentation does not reflect this change yet. In general, unnecessary API breaks like this should be avoided and I have to discuss this with the GLSP team to make sure it doesn't happen in the future.

In the meantime, I believe you can fix this issue by using the `updateRoot` method instead of the `setRoot` method

**Reply by rsoika** [reply]
Yes! Calling `updateRoot` solved the `GLSPServerException`. My model is now correctly loaded and displayed

Thanks a lot!  So now I am back on the road again! :-)

**Reply by rsoika** [reply]
Just to give you some feedback: the new `SourceModelStorage`, `GModelState` and `GModelFactory` is now very clear and easy to implement - and well documented ;-)
